### DISCUSSION TITLE
Add `as` null-forgiving operator

### DISCUSSION THREAD

**Discussion by momvart** [dotnet/csharplang#4796] [Language Ideas] [answered]
Consider the case that we are somehow sure about the concrete type of a class and `as` operator will not generate `null` in the conversion. So we may use the null-forgiving operator to bypass warnings.
```csharp
var x = obj as SomeType;    //x will be SomeType?
var x = (obj as SomeType)!;
```
But IMO it's kind of hard to write and even has made the code a bit ugly.
It would be great if there was a syntactic sugar version of `as` with the null-forgiving operator bound to it: `as!`. Therefore the example would be like:
```csharp
var x = obj as! SomeType;
```
One may argue that it would be ambiguous for the compiler to distinguish between this and the not operator. First, it should be connected with `as` as a special case of `as`, so the parser would distinguish them. By the way, even if not connected, it might be `!SomeType` which is using the not operator with a type that is not valid.

**Comment by yaakov-h** [comment, marked as answer]
Why not use a direct cast, i.e.?:

```csharp
var x = (SomeType)obj;
```

It's faster, too, IIRC.

**Reply by CyrusNajmabadi** [reply]
@yaakov-h is correct here.  The semantics of your `as!` operator seem to be exactly the same as the semantics of the normal `(cast)` operation.

**Reply by momvart** [reply]
I've heard that because of throwing exceptions and some checks cast would be slower. Please, correct me if I'm wrong.

**Reply by CyrusNajmabadi** [reply]
in your scenario it doesn't apply.  As you mentioned, you're " sure about the concrete type".  So no exceptions will happen.

**Reply by yaakov-h** [reply]
`as`/`isinst` (IL) has been well-known for many years to be "slow".

Checking against the latest builds of .NET 5 - I assume you're not on .NET Framework otherwise you're probably not using NRT anyway - this still holds up, with a regular cast being several times faster than using `as`:

```
BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19042.985 (20H2/October2020Update)
AMD Ryzen 9 3900X, 1 CPU, 24 logical and 12 physical cores
.NET SDK=5.0.203
  [Host]     : .NET 5.0.6 (5.0.621.22011), X64 RyuJIT
  DefaultJob : .NET 5.0.6 (5.0.621.22011), X64 RyuJIT


| Method |                    o |      Mean |     Error |    StdDev |
|------- |--------------------- |----------:|----------:|----------:|
|     As |                    ? | 1.6269 ns | 0.0187 ns | 0.0165 ns |
|   Cast |                    ? | 0.2140 ns | 0.0102 ns | 0.0085 ns |
|     As | Class(...)eType [34] | 1.3764 ns | 0.0098 ns | 0.0087 ns |
|   Cast | Class(...)eType [34] | 0.4912 ns | 0.0179 ns | 0.0159 ns |
```

(First two methods are where the input is null, second two are matching concrete type.)

**Reply by momvart** [reply]
@yaakov-h @CyrusNajmabadi 
Thanks for your clarifications. I was mistaken basically!

**Reply by timcassell** [reply]
Also using `as` is a code smell in most cases, a direct cast should almost always be used instead. There are very few cases where using `as` makes sense, especially since the introduction of the `is` pattern match.

**Reply by AartBluestoke** [reply]
> I've heard that because of throwing exceptions and some checks cast would be slower. Please, correct me if I'm wrong.

not throwing an exception from inside a try block is (in general) a 0 cost action, including establishing the try region. 
(i said 'in general' because there are a few corner cases to do with jit code optimisation, and the extra code size of the exception handler, but the IL executed on the no-exception path will be the same.)
